torch/autograd/gradcheck.py

```diff
@@ -285,6 +285,80 @@ def get_numerical_jacobian_wrt_specific_input(fn, input_idx, inputs, outputs, ep
         jacobian_cols[d_idx] = _compute_numerical_jvps_wrt_specific_input(jvp_fn, eps, x.is_complex())
     return _combine_jacobian_cols(jacobian_cols, outputs, input, input.numel())
 
+def _get_analytical_jacobian_forward_ad(fn, inputs, outputs, *, all_u=None) -> Tuple[Tuple[torch.Tensor, ...], ...]:
+    """Computes the analytical Jacobian using forward mode AD of `fn(inputs)` with respect
+    to `target`. Returns M * N Jacobians where N is the number of tensors in target that require grad and
+    M is the number of non-integral outputs.
+    Contrary to other functions here, this function requires "inputs" to actually be used by the function.
+    The computed value is expected to be wrong if the function captures the inputs by side effect instead of
+    using the passed ones (many torch.nn tests do this).
+
+    Args:
+        fn: the function to compute the jacobian for
+        inputs: inputs to `fn`
+        outputs: provide precomputed outputs to avoid one extra invocation of fn
+        all_u (optional): if provided, the Jacobian will be right multiplied with this vector
+
+    Returns:
+        A list of M N-tuples of tensors
```

Comment:
nit: type signature says Tuple[Tuple[Tensor, ...], ...], should this be "a tuple of M N-tuples"?